### PR TITLE
AutoDelta implementation.

### PR DESCRIPTION
**Pull request by xidulu**
AutoGuide for MAP estimation,

### CODE REVIEW

File: numpyro/infer/autoguide.py [hunk 1]
```diff
@@ -261,6 +263,95 @@ def quantiles(self, params, quantiles):
...
+        super().__init__(model, prefix=prefix, init_loc_fn=init_loc_fn, create_plates=create_plates)
+
+    def _setup_prototype(self, *args, **kwargs):
+        super()._setup_prototype(*args, **kwargs)
+
+        for name, site in self.prototype_trace.items():
+            if site["type"] != "sample" or site["is_observed"]:
+                continue
+
+            event_dim = site["fn"].event_dim + jnp.ndim(self._init_locs[name]) - jnp.ndim(site["value"])
```
Comment (fehiepsi) [inline numpyro/infer/autoguide.py:301]: nit: this can be simplified to `event_dim = site["fn"].event_dim` because we are working on constrained domain (this needs the line `self._init_locs = ...` in the above comment).

File: numpyro/infer/autoguide.py [hunk 2]
```diff
@@ -261,6 +263,95 @@ def quantiles(self, params, quantiles):
...
+                result[name] = numpyro.sample(name, site_fn)
+
+        return result
+
+    def sample_posterior(self, rng_key, params, sample_shape=()):
+        locs = {k: params["{}_{}_loc".format(k, self.prefix)] for k in self._init_locs}
+        with handlers.seed(rng_seed=rng_key):
+            latent_samples = {}
+            for k in locs:
+                latent_samples[k] = numpyro.sample(k, dist.Delta(locs[k]).expand_by(sample_shape))
```
Comment (fehiepsi) [inline numpyro/infer/autoguide.py:341]: This is nice but I wonder if we can simply expand each value in `locs` to avoid extra `random.split` under `handlers.seed`.

**Review by fehiepsi — COMMENTED** [review]
Very clean! Thanks a lot, @xidulu! Could you also add an entry to [docs](https://github.com/pyro-ppl/numpyro/blob/master/docs/source/autoguide.rst). I just have some comments below.

File: numpyro/infer/autoguide.py [hunk 3]
```diff
@@ -261,6 +263,95 @@ def quantiles(self, params, quantiles):
...
+        automatically as usual. This is useful for data subsampling.
+    """
+    def __init__(self, model, *, prefix='auto', init_loc_fn=init_to_median,
+                 create_plates=None):
+        self.init_loc_fn = init_loc_fn
+        self._event_dims = {}
+        super().__init__(model, prefix=prefix, init_loc_fn=init_loc_fn, create_plates=create_plates)
+
+    def _setup_prototype(self, *args, **kwargs):
+        super()._setup_prototype(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (xidulu) [reply numpyro/infer/autoguide.py:295]: Oh, yeah, I totally forgot about the constraint.  = =

File: numpyro/infer/autoguide.py (hunk 2 above)
Comment (xidulu) [reply numpyro/infer/autoguide.py:341]: My implementation follows the sample_posterior() from AutoNormal,

Do you mean, directly call apply reshape on locs[k] instead of using expand_by ?

Comment (fehiepsi) [reply numpyro/infer/autoguide.py:341]: I think we can simply return: `{k: jnp.broadcast_to(sample_shape + jnp.shape(v)) for k, v in locs.items}`. What do you think?

Comment (xidulu) [reply numpyro/infer/autoguide.py:341]: Oh I got it, does it mean ```with handlers.seed(rng_seed=rng_key):```  could just be removed? Since delta distribution is deterministic.

Comment (fehiepsi) [reply numpyro/infer/autoguide.py:341]: Yes, we don't need rng_key here.

File: numpyro/infer/autoguide.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (xidulu) [reply numpyro/infer/autoguide.py:295]: @fehiepsi 
How about 
```
self._init_locs = {k: v for k, v in self._postprocess_fn(self._init_locs).items()}
```
Will it have the same effect?

[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/infer/autoguide.py:295]: Sorry, I should make it clearer. postprocess might return deterministic sites, which we need to filter them out.

[the remark this answers is not in the document]
Comment (xidulu) [reply numpyro/infer/autoguide.py:295]: @fehiepsi 
Now the deterministic sites should be filtered out.
https://github.com/pyro-ppl/numpyro/pull/816/files#diff-84384c2f732a05bdda11794d2be295bd6e174d644562a1c565260d289464e95aR297

[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/infer/autoguide.py:295]: Thanks! It is a bug in my above code. ^^

**Review by fehiepsi — APPROVED** [review]
LGTM, thanks, @xidulu! Please let me know if this is already to merge. :)

**Comment by xidulu** [comment]
> LGTM, thanks, @xidulu! Please let me know if this is already to merge. :)

Go ahead plz.
